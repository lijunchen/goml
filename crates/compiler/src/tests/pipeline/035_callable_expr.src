fn double(x: int) -> int {
    x * 2
}

fn increment(x: int) -> int {
    x + 1
}

fn make(adder: int) -> (int) -> int {
    |value: int| value + adder
}

fn chooser(flag: bool) -> (int) -> int {
    if flag { double } else { increment }
}

fn main() {
    let result_direct = chooser(true)(3) in
    let chosen = chooser(false) in
    let result_from_var = chosen(4) in
    let apply = |f: (int) -> int, value: int| f(value) in
    let closure_call = apply(make(5), 1) in
    let composed = make(6)(2) in
    let generator = make(7) in
    let from_generator = generator(3) in
    let printer = string_println in
    let _ = printer(int_to_string(result_direct)) in
    let _ = printer(int_to_string(result_from_var)) in
    let _ = printer(int_to_string(closure_call)) in
    let _ = printer(int_to_string(composed)) in
    let _ = printer(int_to_string(from_generator)) in
    ()
}
