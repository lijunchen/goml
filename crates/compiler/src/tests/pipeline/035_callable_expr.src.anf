fn double(x/0: int) -> int {
  int_mul(x/0, 2)
}

fn increment(x/1: int) -> int {
  int_add(x/1, 1)
}

fn make(adder/2: int) -> (int) -> int {
  ClosureEnv0 { adder_2_0: adder/2 }
}

fn chooser(flag/4: bool) -> (int) -> int {
  if flag/4 {
    double
  } else {
    increment
  }
}

fn main() -> unit {
  let t9 = chooser(true) in
  let result_direct/5 = t9(3) in
  let chosen/6 = chooser(false) in
  let result_from_var/7 = chosen/6(4) in
  let apply/10 = ClosureEnv1 {} in
  let t10 = make(5) in
  let closure_call/11 = __closure_apply7(apply/10, t10, 1) in
  let t11 = make(6) in
  let composed/12 = t11(2) in
  let generator/13 = make(7) in
  let from_generator/14 = generator/13(3) in
  let printer/15 = string_println in
  let t12 = int_to_string(result_direct/5) in
  let mtmp0 = printer/15(t12) in
  let t13 = int_to_string(result_from_var/7) in
  let mtmp1 = printer/15(t13) in
  let t14 = int_to_string(closure_call/11) in
  let mtmp2 = printer/15(t14) in
  let t15 = int_to_string(composed/12) in
  let mtmp3 = printer/15(t15) in
  let t16 = int_to_string(from_generator/14) in
  let mtmp4 = printer/15(t16) in
  ()
}

fn __closure_apply5(env6: ClosureEnv0, value/3: int) -> int {
  let adder/2 = ClosureEnv0.adder_2_0(env6) in
  int_add(value/3, adder/2)
}

fn __closure_apply7(env8: ClosureEnv1, f/8: (int) -> int, value/9: int) -> int {
  f/8(value/9)
}