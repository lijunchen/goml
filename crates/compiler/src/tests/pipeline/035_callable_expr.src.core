fn double(x/0: int) -> int {
  int_mul(x/0, 2)
}

fn increment(x/1: int) -> int {
  int_add(x/1, 1)
}

fn make(adder/2: int) -> (int) -> int {
  |value/3: int| => int_add(value/3, adder/2)
}

fn chooser(flag/4: bool) -> (int) -> int {
  if flag/4 {
      double
  } else {
      increment
  }
}

fn main() -> unit {
  let result_direct/5 = chooser(true)(3) in
  let chosen/6 = chooser(false) in
  let result_from_var/7 = chosen/6(4) in
  let apply/10 = |f/8: (int) -> int, value/9: int| => f/8(value/9) in
  let closure_call/11 = apply/10(make(5), 1) in
  let composed/12 = make(6)(2) in
  let generator/13 = make(7) in
  let from_generator/14 = generator/13(3) in
  let printer/15 = string_println in
  let mtmp0 = printer/15(int_to_string(result_direct/5)) in
  let mtmp1 = printer/15(int_to_string(result_from_var/7)) in
  let mtmp2 = printer/15(int_to_string(closure_call/11)) in
  let mtmp3 = printer/15(int_to_string(composed/12)) in
  let mtmp4 = printer/15(int_to_string(from_generator/14)) in
  ()
}