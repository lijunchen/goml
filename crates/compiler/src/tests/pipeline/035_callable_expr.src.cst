FILE@0..885
  FN@0..40
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..9 "double"
    PARAM_LIST@9..18
      LParen@9..10 "("
      PARAM@10..16
        Lident@10..11 "x"
        Colon@11..12 ":"
        Whitespace@12..13 " "
        TYPE_INT@13..16
          IntKeyword@13..16 "int"
      RParen@16..17 ")"
      Whitespace@17..18 " "
    Arrow@18..20 "->"
    Whitespace@20..21 " "
    TYPE_INT@21..25
      IntKeyword@21..24 "int"
      Whitespace@24..25 " "
    BLOCK@25..40
      LBrace@25..26 "{"
      Whitespace@26..31 "\n    "
      EXPR_BINARY@31..37
        EXPR_LIDENT@31..33
          Lident@31..32 "x"
          Whitespace@32..33 " "
        Star@33..34 "*"
        Whitespace@34..35 " "
        EXPR_INT@35..37
          Int@35..36 "2"
          Whitespace@36..37 "\n"
      RBrace@37..38 "}"
      Whitespace@38..40 "\n\n"
  FN@40..83
    FnKeyword@40..42 "fn"
    Whitespace@42..43 " "
    Lident@43..52 "increment"
    PARAM_LIST@52..61
      LParen@52..53 "("
      PARAM@53..59
        Lident@53..54 "x"
        Colon@54..55 ":"
        Whitespace@55..56 " "
        TYPE_INT@56..59
          IntKeyword@56..59 "int"
      RParen@59..60 ")"
      Whitespace@60..61 " "
    Arrow@61..63 "->"
    Whitespace@63..64 " "
    TYPE_INT@64..68
      IntKeyword@64..67 "int"
      Whitespace@67..68 " "
    BLOCK@68..83
      LBrace@68..69 "{"
      Whitespace@69..74 "\n    "
      EXPR_BINARY@74..80
        EXPR_LIDENT@74..76
          Lident@74..75 "x"
          Whitespace@75..76 " "
        Plus@76..77 "+"
        Whitespace@77..78 " "
        EXPR_INT@78..80
          Int@78..79 "1"
          Whitespace@79..80 "\n"
      RBrace@80..81 "}"
      Whitespace@81..83 "\n\n"
  FN@83..155
    FnKeyword@83..85 "fn"
    Whitespace@85..86 " "
    Lident@86..90 "make"
    PARAM_LIST@90..103
      LParen@90..91 "("
      PARAM@91..101
        Lident@91..96 "adder"
        Colon@96..97 ":"
        Whitespace@97..98 " "
        TYPE_INT@98..101
          IntKeyword@98..101 "int"
      RParen@101..102 ")"
      Whitespace@102..103 " "
    Arrow@103..105 "->"
    Whitespace@105..106 " "
    TYPE_FUNC@106..119
      TYPE_TUPLE@106..112
        TYPE_LIST@106..112
          LParen@106..107 "("
          TYPE_INT@107..110
            IntKeyword@107..110 "int"
          RParen@110..111 ")"
          Whitespace@111..112 " "
      Arrow@112..114 "->"
      Whitespace@114..115 " "
      TYPE_INT@115..119
        IntKeyword@115..118 "int"
        Whitespace@118..119 " "
    BLOCK@119..155
      LBrace@119..120 "{"
      Whitespace@120..125 "\n    "
      EXPR_CLOSURE@125..152
        CLOSURE_PARAM_LIST@125..138
          Pipe@125..126 "|"
          CLOSURE_PARAM@126..136
            Lident@126..131 "value"
            Colon@131..132 ":"
            Whitespace@132..133 " "
            TYPE_INT@133..136
              IntKeyword@133..136 "int"
          Pipe@136..137 "|"
          Whitespace@137..138 " "
        EXPR_CLOSURE_BODY@138..152
          EXPR_BINARY@138..152
            EXPR_LIDENT@138..144
              Lident@138..143 "value"
              Whitespace@143..144 " "
            Plus@144..145 "+"
            Whitespace@145..146 " "
            EXPR_LIDENT@146..152
              Lident@146..151 "adder"
              Whitespace@151..152 "\n"
      RBrace@152..153 "}"
      Whitespace@153..155 "\n\n"
  FN@155..241
    FnKeyword@155..157 "fn"
    Whitespace@157..158 " "
    Lident@158..165 "chooser"
    PARAM_LIST@165..178
      LParen@165..166 "("
      PARAM@166..176
        Lident@166..170 "flag"
        Colon@170..171 ":"
        Whitespace@171..172 " "
        TYPE_BOOL@172..176
          BoolKeyword@172..176 "bool"
      RParen@176..177 ")"
      Whitespace@177..178 " "
    Arrow@178..180 "->"
    Whitespace@180..181 " "
    TYPE_FUNC@181..194
      TYPE_TUPLE@181..187
        TYPE_LIST@181..187
          LParen@181..182 "("
          TYPE_INT@182..185
            IntKeyword@182..185 "int"
          RParen@185..186 ")"
          Whitespace@186..187 " "
      Arrow@187..189 "->"
      Whitespace@189..190 " "
      TYPE_INT@190..194
        IntKeyword@190..193 "int"
        Whitespace@193..194 " "
    BLOCK@194..241
      LBrace@194..195 "{"
      Whitespace@195..200 "\n    "
      EXPR_IF@200..238
        IfKeyword@200..202 "if"
        Whitespace@202..203 " "
        EXPR_IF_COND@203..208
          EXPR_LIDENT@203..208
            Lident@203..207 "flag"
            Whitespace@207..208 " "
        EXPR_IF_THEN@208..219
          BLOCK@208..219
            LBrace@208..209 "{"
            Whitespace@209..210 " "
            EXPR_LIDENT@210..217
              Lident@210..216 "double"
              Whitespace@216..217 " "
            RBrace@217..218 "}"
            Whitespace@218..219 " "
        ElseKeyword@219..223 "else"
        Whitespace@223..224 " "
        EXPR_IF_ELSE@224..238
          BLOCK@224..238
            LBrace@224..225 "{"
            Whitespace@225..226 " "
            EXPR_LIDENT@226..236
              Lident@226..235 "increment"
              Whitespace@235..236 " "
            RBrace@236..237 "}"
            Whitespace@237..238 "\n"
      RBrace@238..239 "}"
      Whitespace@239..241 "\n\n"
  FN@241..885
    FnKeyword@241..243 "fn"
    Whitespace@243..244 " "
    Lident@244..248 "main"
    PARAM_LIST@248..251
      LParen@248..249 "("
      RParen@249..250 ")"
      Whitespace@250..251 " "
    BLOCK@251..885
      LBrace@251..252 "{"
      Whitespace@252..257 "\n    "
      EXPR_LET@257..883
        LetKeyword@257..260 "let"
        Whitespace@260..261 " "
        PATTERN_VARIABLE@261..275
          Lident@261..274 "result_direct"
          Whitespace@274..275 " "
        Eq@275..276 "="
        Whitespace@276..277 " "
        EXPR_LET_VALUE@277..294
          EXPR_CALL@277..294
            EXPR_CALL@277..290
              EXPR_LIDENT@277..284
                Lident@277..284 "chooser"
              ARG_LIST@284..290
                LParen@284..285 "("
                ARG@285..289
                  EXPR_BOOL@285..289
                    TrueKeyword@285..289 "true"
                RParen@289..290 ")"
            ARG_LIST@290..294
              LParen@290..291 "("
              ARG@291..292
                EXPR_INT@291..292
                  Int@291..292 "3"
              RParen@292..293 ")"
              Whitespace@293..294 " "
        InKeyword@294..296 "in"
        Whitespace@296..301 "\n    "
        EXPR_LET_BODY@301..883
          EXPR_LET@301..883
            LetKeyword@301..304 "let"
            Whitespace@304..305 " "
            PATTERN_VARIABLE@305..312
              Lident@305..311 "chosen"
              Whitespace@311..312 " "
            Eq@312..313 "="
            Whitespace@313..314 " "
            EXPR_LET_VALUE@314..329
              EXPR_CALL@314..329
                EXPR_LIDENT@314..321
                  Lident@314..321 "chooser"
                ARG_LIST@321..329
                  LParen@321..322 "("
                  ARG@322..327
                    EXPR_BOOL@322..327
                      FalseKeyword@322..327 "false"
                  RParen@327..328 ")"
                  Whitespace@328..329 " "
            InKeyword@329..331 "in"
            Whitespace@331..336 "\n    "
            EXPR_LET_BODY@336..883
              EXPR_LET@336..883
                LetKeyword@336..339 "let"
                Whitespace@339..340 " "
                PATTERN_VARIABLE@340..356
                  Lident@340..355 "result_from_var"
                  Whitespace@355..356 " "
                Eq@356..357 "="
                Whitespace@357..358 " "
                EXPR_LET_VALUE@358..368
                  EXPR_CALL@358..368
                    EXPR_LIDENT@358..364
                      Lident@358..364 "chosen"
                    ARG_LIST@364..368
                      LParen@364..365 "("
                      ARG@365..366
                        EXPR_INT@365..366
                          Int@365..366 "4"
                      RParen@366..367 ")"
                      Whitespace@367..368 " "
                InKeyword@368..370 "in"
                Whitespace@370..375 "\n    "
                EXPR_LET_BODY@375..883
                  EXPR_LET@375..883
                    LetKeyword@375..378 "let"
                    Whitespace@378..379 " "
                    PATTERN_VARIABLE@379..385
                      Lident@379..384 "apply"
                      Whitespace@384..385 " "
                    Eq@385..386 "="
                    Whitespace@386..387 " "
                    EXPR_LET_VALUE@387..426
                      EXPR_CLOSURE@387..426
                        CLOSURE_PARAM_LIST@387..417
                          Pipe@387..388 "|"
                          CLOSURE_PARAM@388..403
                            Lident@388..389 "f"
                            Colon@389..390 ":"
                            Whitespace@390..391 " "
                            TYPE_FUNC@391..403
                              TYPE_TUPLE@391..397
                                TYPE_LIST@391..397
                                  LParen@391..392 "("
                                  TYPE_INT@392..395
                                    IntKeyword@392..395 "int"
                                  RParen@395..396 ")"
                                  Whitespace@396..397 " "
                              Arrow@397..399 "->"
                              Whitespace@399..400 " "
                              TYPE_INT@400..403
                                IntKeyword@400..403 "int"
                          Comma@403..404 ","
                          Whitespace@404..405 " "
                          CLOSURE_PARAM@405..415
                            Lident@405..410 "value"
                            Colon@410..411 ":"
                            Whitespace@411..412 " "
                            TYPE_INT@412..415
                              IntKeyword@412..415 "int"
                          Pipe@415..416 "|"
                          Whitespace@416..417 " "
                        EXPR_CLOSURE_BODY@417..426
                          EXPR_CALL@417..426
                            EXPR_LIDENT@417..418
                              Lident@417..418 "f"
                            ARG_LIST@418..426
                              LParen@418..419 "("
                              ARG@419..424
                                EXPR_LIDENT@419..424
                                  Lident@419..424 "value"
                              RParen@424..425 ")"
                              Whitespace@425..426 " "
                    InKeyword@426..428 "in"
                    Whitespace@428..433 "\n    "
                    EXPR_LET_BODY@433..883
                      EXPR_LET@433..883
                        LetKeyword@433..436 "let"
                        Whitespace@436..437 " "
                        PATTERN_VARIABLE@437..450
                          Lident@437..449 "closure_call"
                          Whitespace@449..450 " "
                        Eq@450..451 "="
                        Whitespace@451..452 " "
                        EXPR_LET_VALUE@452..470
                          EXPR_CALL@452..470
                            EXPR_LIDENT@452..457
                              Lident@452..457 "apply"
                            ARG_LIST@457..470
                              LParen@457..458 "("
                              ARG@458..467
                                EXPR_CALL@458..465
                                  EXPR_LIDENT@458..462
                                    Lident@458..462 "make"
                                  ARG_LIST@462..465
                                    LParen@462..463 "("
                                    ARG@463..464
                                      EXPR_INT@463..464
                                        Int@463..464 "5"
                                    RParen@464..465 ")"
                                Comma@465..466 ","
                                Whitespace@466..467 " "
                              ARG@467..468
                                EXPR_INT@467..468
                                  Int@467..468 "1"
                              RParen@468..469 ")"
                              Whitespace@469..470 " "
                        InKeyword@470..472 "in"
                        Whitespace@472..477 "\n    "
                        EXPR_LET_BODY@477..883
                          EXPR_LET@477..883
                            LetKeyword@477..480 "let"
                            Whitespace@480..481 " "
                            PATTERN_VARIABLE@481..490
                              Lident@481..489 "composed"
                              Whitespace@489..490 " "
                            Eq@490..491 "="
                            Whitespace@491..492 " "
                            EXPR_LET_VALUE@492..503
                              EXPR_CALL@492..503
                                EXPR_CALL@492..499
                                  EXPR_LIDENT@492..496
                                    Lident@492..496 "make"
                                  ARG_LIST@496..499
                                    LParen@496..497 "("
                                    ARG@497..498
                                      EXPR_INT@497..498
                                        Int@497..498 "6"
                                    RParen@498..499 ")"
                                ARG_LIST@499..503
                                  LParen@499..500 "("
                                  ARG@500..501
                                    EXPR_INT@500..501
                                      Int@500..501 "2"
                                  RParen@501..502 ")"
                                  Whitespace@502..503 " "
                            InKeyword@503..505 "in"
                            Whitespace@505..510 "\n    "
                            EXPR_LET_BODY@510..883
                              EXPR_LET@510..883
                                LetKeyword@510..513 "let"
                                Whitespace@513..514 " "
                                PATTERN_VARIABLE@514..524
                                  Lident@514..523 "generator"
                                  Whitespace@523..524 " "
                                Eq@524..525 "="
                                Whitespace@525..526 " "
                                EXPR_LET_VALUE@526..534
                                  EXPR_CALL@526..534
                                    EXPR_LIDENT@526..530
                                      Lident@526..530 "make"
                                    ARG_LIST@530..534
                                      LParen@530..531 "("
                                      ARG@531..532
                                        EXPR_INT@531..532
                                          Int@531..532 "7"
                                      RParen@532..533 ")"
                                      Whitespace@533..534 " "
                                InKeyword@534..536 "in"
                                Whitespace@536..541 "\n    "
                                EXPR_LET_BODY@541..883
                                  EXPR_LET@541..883
                                    LetKeyword@541..544 "let"
                                    Whitespace@544..545 " "
                                    PATTERN_VARIABLE@545..560
                                      Lident@545..559 "from_generator"
                                      Whitespace@559..560 " "
                                    Eq@560..561 "="
                                    Whitespace@561..562 " "
                                    EXPR_LET_VALUE@562..575
                                      EXPR_CALL@562..575
                                        EXPR_LIDENT@562..571
                                          Lident@562..571 "generator"
                                        ARG_LIST@571..575
                                          LParen@571..572 "("
                                          ARG@572..573
                                            EXPR_INT@572..573
                                              Int@572..573 "3"
                                          RParen@573..574 ")"
                                          Whitespace@574..575 " "
                                    InKeyword@575..577 "in"
                                    Whitespace@577..582 "\n    "
                                    EXPR_LET_BODY@582..883
                                      EXPR_LET@582..883
                                        LetKeyword@582..585 "let"
                                        Whitespace@585..586 " "
                                        PATTERN_VARIABLE@586..594
                                          Lident@586..593 "printer"
                                          Whitespace@593..594 " "
                                        Eq@594..595 "="
                                        Whitespace@595..596 " "
                                        EXPR_LET_VALUE@596..611
                                          EXPR_LIDENT@596..611
                                            Lident@596..610 "string_println"
                                            Whitespace@610..611 " "
                                        InKeyword@611..613 "in"
                                        Whitespace@613..618 "\n    "
                                        EXPR_LET_BODY@618..883
                                          EXPR_LET@618..883
                                            LetKeyword@618..621 "let"
                                            Whitespace@621..622 " "
                                            PATTERN_WILDCARD@622..624
                                              WildcardKeyword@622..623 "_"
                                              Whitespace@623..624 " "
                                            Eq@624..625 "="
                                            Whitespace@625..626 " "
                                            EXPR_LET_VALUE@626..664
                                              EXPR_CALL@626..664
                                                EXPR_LIDENT@626..633
                                                  Lident@626..633 "printer"
                                                ARG_LIST@633..664
                                                  LParen@633..634 "("
                                                  ARG@634..662
                                                    EXPR_CALL@634..662
                                                      EXPR_LIDENT@634..647
                                                        Lident@634..647 "int_to_string"
                                                      ARG_LIST@647..662
                                                        LParen@647..648 "("
                                                        ARG@648..661
                                                          EXPR_LIDENT@648..661
                                                            Lident@648..661 "result_direct"
                                                        RParen@661..662 ")"
                                                  RParen@662..663 ")"
                                                  Whitespace@663..664 " "
                                            InKeyword@664..666 "in"
                                            Whitespace@666..671 "\n    "
                                            EXPR_LET_BODY@671..883
                                              EXPR_LET@671..883
                                                LetKeyword@671..674 "let"
                                                Whitespace@674..675 " "
                                                PATTERN_WILDCARD@675..677
                                                  WildcardKeyword@675..676 "_"
                                                  Whitespace@676..677 " "
                                                Eq@677..678 "="
                                                Whitespace@678..679 " "
                                                EXPR_LET_VALUE@679..719
                                                  EXPR_CALL@679..719
                                                    EXPR_LIDENT@679..686
                                                      Lident@679..686 "printer"
                                                    ARG_LIST@686..719
                                                      LParen@686..687 "("
                                                      ARG@687..717
                                                        EXPR_CALL@687..717
                                                          EXPR_LIDENT@687..700
                                                            Lident@687..700 "int_to_string"
                                                          ARG_LIST@700..717
                                                            LParen@700..701 "("
                                                            ARG@701..716
                                                              EXPR_LIDENT@701..716
                                                                Lident@701..716 "result_from_var"
                                                            RParen@716..717 ")"
                                                      RParen@717..718 ")"
                                                      Whitespace@718..719 " "
                                                InKeyword@719..721 "in"
                                                Whitespace@721..726 "\n    "
                                                EXPR_LET_BODY@726..883
                                                  EXPR_LET@726..883
                                                    LetKeyword@726..729 "let"
                                                    Whitespace@729..730 " "
                                                    PATTERN_WILDCARD@730..732
                                                      WildcardKeyword@730..731 "_"
                                                      Whitespace@731..732 " "
                                                    Eq@732..733 "="
                                                    Whitespace@733..734 " "
                                                    EXPR_LET_VALUE@734..771
                                                      EXPR_CALL@734..771
                                                        EXPR_LIDENT@734..741
                                                          Lident@734..741 "printer"
                                                        ARG_LIST@741..771
                                                          LParen@741..742 "("
                                                          ARG@742..769
                                                            EXPR_CALL@742..769
                                                              EXPR_LIDENT@742..755
                                                                Lident@742..755 "int_to_string"
                                                              ARG_LIST@755..769
                                                                LParen@755..756 "("
                                                                ARG@756..768
                                                                  EXPR_LIDENT@756..768
                                                                    Lident@756..768 "closure_call"
                                                                RParen@768..769 ")"
                                                          RParen@769..770 ")"
                                                          Whitespace@770..771 " "
                                                    InKeyword@771..773 "in"
                                                    Whitespace@773..778 "\n    "
                                                    EXPR_LET_BODY@778..883
                                                      EXPR_LET@778..883
                                                        LetKeyword@778..781 "let"
                                                        Whitespace@781..782 " "
                                                        PATTERN_WILDCARD@782..784
                                                          WildcardKeyword@782..783 "_"
                                                          Whitespace@783..784 " "
                                                        Eq@784..785 "="
                                                        Whitespace@785..786 " "
                                                        EXPR_LET_VALUE@786..819
                                                          EXPR_CALL@786..819
                                                            EXPR_LIDENT@786..793
                                                              Lident@786..793 "printer"
                                                            ARG_LIST@793..819
                                                              LParen@793..794 "("
                                                              ARG@794..817
                                                                EXPR_CALL@794..817
                                                                  EXPR_LIDENT@794..807
                                                                    Lident@794..807 "int_to_string"
                                                                  ARG_LIST@807..817
                                                                    LParen@807..808 "("
                                                                    ARG@808..816
                                                                      EXPR_LIDENT@808..816
                                                                        Lident@808..816 "composed"
                                                                    RParen@816..817 ")"
                                                              RParen@817..818 ")"
                                                              Whitespace@818..819 " "
                                                        InKeyword@819..821 "in"
                                                        Whitespace@821..826 "\n    "
                                                        EXPR_LET_BODY@826..883
                                                          EXPR_LET@826..883
                                                            LetKeyword@826..829 "let"
                                                            Whitespace@829..830 " "
                                                            PATTERN_WILDCARD@830..832
                                                              WildcardKeyword@830..831 "_"
                                                              Whitespace@831..832 " "
                                                            Eq@832..833 "="
                                                            Whitespace@833..834 " "
                                                            EXPR_LET_VALUE@834..873
                                                              EXPR_CALL@834..873
                                                                EXPR_LIDENT@834..841
                                                                  Lident@834..841 "printer"
                                                                ARG_LIST@841..873
                                                                  LParen@841..842 "("
                                                                  ARG@842..871
                                                                    EXPR_CALL@842..871
                                                                      EXPR_LIDENT@842..855
                                                                        Lident@842..855 "int_to_string"
                                                                      ARG_LIST@855..871
                                                                        LParen@855..856 "("
                                                                        ARG@856..870
                                                                          EXPR_LIDENT@856..870
                                                                            Lident@856..870 "from_generator"
                                                                        RParen@870..871 ")"
                                                                  RParen@871..872 ")"
                                                                  Whitespace@872..873 " "
                                                            InKeyword@873..875 "in"
                                                            Whitespace@875..880 "\n    "
                                                            EXPR_LET_BODY@880..883
                                                              EXPR_UNIT@880..883
                                                                LParen@880..881 "("
                                                                RParen@881..882 ")"
                                                                Whitespace@882..883 "\n"
      RBrace@883..884 "}"
      Whitespace@884..885 "\n"