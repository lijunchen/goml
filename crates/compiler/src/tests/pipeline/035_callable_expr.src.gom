package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_mul(x int, y int) int {
    return x * y
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

type ClosureEnv0 struct {
    adder_2_0 int
}

type ClosureEnv1 struct {}

func double(x__0 int) int {
    var ret17 int
    ret17 = int_mul(x__0, 2)
    return ret17
}

func increment(x__1 int) int {
    var ret18 int
    ret18 = int_add(x__1, 1)
    return ret18
}

func make(adder__2 int) func(int) int {
    var ret19 func(int) int
    ret19 = ClosureEnv0{
        adder_2_0: adder__2,
    }
    return ret19
}

func chooser(flag__4 bool) func(int) int {
    var ret20 func(int) int
    if flag__4 {
        ret20 = double
    } else {
        ret20 = increment
    }
    return ret20
}

func main0() struct{} {
    var ret21 struct{}
    var t9 func(int) int = chooser(true)
    var result_direct__5 int = t9(3)
    var chosen__6 func(int) int = chooser(false)
    var result_from_var__7 int = chosen__6(4)
    var apply__10 ClosureEnv1 = ClosureEnv1{}
    var t10 func(int) int = make(5)
    var closure_call__11 int = __closure_apply7(apply__10, t10, 1)
    var t11 func(int) int = make(6)
    var composed__12 int = t11(2)
    var generator__13 func(int) int = make(7)
    var from_generator__14 int = generator__13(3)
    var printer__15 func(string) struct{} = string_println
    var t12 string = int_to_string(result_direct__5)
    printer__15(t12)
    var t13 string = int_to_string(result_from_var__7)
    printer__15(t13)
    var t14 string = int_to_string(closure_call__11)
    printer__15(t14)
    var t15 string = int_to_string(composed__12)
    printer__15(t15)
    var t16 string = int_to_string(from_generator__14)
    printer__15(t16)
    ret21 = struct{}{}
    return ret21
}

func __closure_apply7(env8 ClosureEnv1, f__8 func(int) int, value__9 int) int {
    var ret23 int
    ret23 = f__8(value__9)
    return ret23
}

func main() {
    main0()
}
