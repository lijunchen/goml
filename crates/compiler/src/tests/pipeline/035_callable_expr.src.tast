fn double(x/0: int) -> int {
  (x/0 : int) * 2
}

fn increment(x/1: int) -> int {
  (x/1 : int) + 1
}

fn make(adder/2: int) -> (int) -> int {
  |value/3: int| => (value/3 : int) + (adder/2 : int)
}

fn chooser(flag/4: bool) -> (int) -> int {
  if (flag/4 : bool) {
      (double : (int) -> int)
  } else {
      (increment : (int) -> int)
  }
}

fn main() -> unit {
  let result_direct/5: int = (chooser : (bool) -> (int) -> int)(true)(3) in
  let chosen/6: (int) -> int = (chooser : (bool) -> (int) -> int)(false) in
  let result_from_var/7: int = (chosen/6 : (int) -> int)(4) in
  let apply/10: ((int) -> int, int) -> int = |f/8: (int) -> int, value/9: int| => (f/8 : (int) -> int)((value/9 : int)) in
  let closure_call/11: int = (apply/10 : ((int) -> int, int) -> int)((make : (int) -> (int) -> int)(5), 1) in
  let composed/12: int = (make : (int) -> (int) -> int)(6)(2) in
  let generator/13: (int) -> int = (make : (int) -> (int) -> int)(7) in
  let from_generator/14: int = (generator/13 : (int) -> int)(3) in
  let printer/15: (string) -> unit = (string_println : (string) -> unit) in
  let _ : unit = (printer/15 : (string) -> unit)((int_to_string : (int) -> string)((result_direct/5 : int))) in
  let _ : unit = (printer/15 : (string) -> unit)((int_to_string : (int) -> string)((result_from_var/7 : int))) in
  let _ : unit = (printer/15 : (string) -> unit)((int_to_string : (int) -> string)((closure_call/11 : int))) in
  let _ : unit = (printer/15 : (string) -> unit)((int_to_string : (int) -> string)((composed/12 : int))) in
  let _ : unit = (printer/15 : (string) -> unit)((int_to_string : (int) -> string)((from_generator/14 : int))) in
  ()
}