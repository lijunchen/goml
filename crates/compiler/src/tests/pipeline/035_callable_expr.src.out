# command-line-arguments
./main.go:44:13: cannot use ClosureEnv0{â€¦} (value of struct type ClosureEnv0) as func(int) int value in assignment
